### PR TITLE
refactor(executor): refactor ArrayImpl::Bool for cheap cloning

### PR DESCRIPTION
**Pull request by kwannoel**
Part of #464, implements solution for `ArrayImpl::Bool`

Signed-off-by: Noel Kwan [noelkwan1998@gmail.com](mailto:noelkwan1998@gmail.com)

### CODE REVIEW

**Comment by kwannoel** [comment]
Will clean-up commits later -- this is just a draft to demo approach for: https://github.com/risinglightdb/risinglight/issues/464

File: src/array/mod.rs [hunk 1]
```diff
@@ -234,7 +243,8 @@ macro_rules! impl_from {
             /// Implement `AbcArray -> ArrayImpl`
             impl From<$AbcArray> for ArrayImpl {
                 fn from(array: $AbcArray) -> Self {
-                    Self::$Abc(array)
+                    // Self::$Abc(array)
```
Comment (xxchan) [inline src/array/mod.rs:246]: Can we `Self::$Abc(Arc::new(array))` here? Then no need to add `new_bool`

Comment (skyzh) [reply src/array/mod.rs:246]: `Self::$Abc(array.into())` might be better. By using `.into()`, we can have both `Arc` and the original representation in such intermediate state.

File: src/array/mod.rs [hunk 2]
```diff
@@ -244,7 +254,8 @@ macro_rules! impl_from {
 
                 fn try_from(array: ArrayImpl) -> Result<Self, Self::Error> {
                     match array {
-                        ArrayImpl::$Abc(array) => Ok(array),
+                        // ArrayImpl::$Abc(array) => Ok(array),
+                        ArrayImpl::$Abc(array) => loop {},
```
Comment (skyzh) [inline src/array/mod.rs:258]: I think this function should be implemented on `Arc<Array>` then. `ArrayImpl::Bool` can now only be turned into an `Arc<BoolArray>`.

File: src/executor/evaluator.rs [hunk 3]
```diff
@@ -45,7 +45,7 @@ impl BoundExpr {
             }
             BoundExpr::IsNull(expr) => {
                 let array = expr.expr.eval(chunk)?;
-                Ok(ArrayImpl::Bool(
+                Ok(ArrayImpl::new_bool(
```
Comment (skyzh) [inline src/executor/evaluator.rs:48]: ArrayImpl::Bool(<the original BoolArray>.into())

**Review by skyzh — COMMENTED** [review]
Generally LGTM, and you are on the right direction! Thanks for this PR.

**Comment by skyzh** [comment]
Well, I've tried doing this refactoring on my local branch. It already took me ~20min to doing the refactor (which produces a ~400LoC's diff), and I think I'll take another 20min to complete the whole thing... Some suggestions:

Use `.into()` to transform a AbcArray into Arc<AbcArray>. For example,

```rust
            impl From<$AbcArray> for ArrayImpl {
                fn from(array: $AbcArray) -> Self {
                    Self::$Abc(array.into())
                }
            }
```

Use the `paste` crate to generate boilerplate code:

```rust
            $(
                paste! {
                    /// Create a new array of the corresponding type
                    pub fn [<new_ $abc>](array: $AbcArray) -> Self {
                        ArrayImpl::$Abc(array.into())
                    }
                }
            )*
```

Sometimes the compiler cannot infer type, especially at:

```
ArrayImpl::Bool([v].into_iter().collect() /* cannot determine what to collect! */),
```

We can use `new_bool` to help the compiler to determine type:

```rust
ArrayImpl::new_bool([v].into_iter().collect()),
```

Anyway, nice try for this PR! I think you have already done a good job and on the right direction. You may continue to finish the `BoolArray` part, and if you're interested, benchmark `select bool_column from table`. (However, from my perspective, I think blob array and string array can have more demonstrating benchmark result, as their clone are much more expensive than other primitive arrays.) After the BoolArray change get merged, I can help refactor all other places.

**Comment by skyzh** [comment]
Also please use `panic!("not supported")` or `todo!()` for code that are not implemented... Otherwise CI will spin for too long time :)

image

**Comment by kwannoel** [comment]
> Also please use `panic!("not supported")` or `todo!()` for code that are not implemented... Otherwise CI will spin for too long time :)
> 
> image

Really sorry about that, will do so.

File: src/executor/evaluator.rs (hunk 3 above)
Comment (kwannoel) [reply src/executor/evaluator.rs:48]: Leaving this as it uses `collect` to construct array.

File: src/array/mod.rs (hunk 1 above)
Comment (kwannoel) [reply src/array/mod.rs]: Great suggestion, thanks!

File: src/array/mod.rs (hunk 2 above)
Comment (kwannoel) [reply src/array/mod.rs]: Done, thanks for suggesting this.

**Review by skyzh — APPROVED** [review]
LGTM, thanks! Are you going to add more commits? If not, I'd be happy to merge this.

**Comment by kwannoel** [comment]
> LGTM, thanks! Are you going to add more commits? If not, I'd be happy to merge this.

Nope not adding anymore changes - but feel free to let me know if you'd like anymore help with refactoring the rest of `ArrayImpl`'s members :) I can do so in separate PRs

**Pull request merged**